### PR TITLE
[RFC][TableGen] Restructure TableGen Source

### LINKED ISSUE #80647: [TableGen] Fix llvm-tblgen library structure
llvm/utils/TableGen has a weird library structure, and it's not the first time a seemingly innocent change triggers linker errors in some build bots. I'm often unable to reproduce such errors locally, so I wonder if they only happen in some configurations, or when using gcc?

In any case, that library is due for a clean-up I believe. Helpers are mixed with backends, some files are ad-hoc placed into other targets because they have unit tests, etc. It may be a good idea to simply have two folders: one for the backends, and one for all helpers associated with the backends.

I had to revert https://github.com/llvm/llvm-project/commit/cfa0833ccc7450a322e709583e894e4c96ce682e because of these issues. This should re-land when fixed.

### PR DESCRIPTION
**Pull request by Pierre-vh**
Lately, while trying to move as much code out of the GlobalISel Combiner Backend, I've been hitting more and more linker issues. I've had to revert one of my recent patches because a buildbot just wouldn't let it slide due to weirdness in how the dependencies were declared. It seems like the library structure was due for an overhaul, so I created #80647

I'm proposing this refactor of the llvm-tblgen source into:
- a "Basic" library, which contains the bare minimum utilities to build `llvm-min-tablegen`
- a "Common" library which contains all of the helpers for TableGen backends. Such helpers can be shared by more than one backend, and even unit tested (e.g. CodeExpander is, maybe we can add more over time)

The backends remain in place and just make use of the new include paths.

I did a directory structure that I hope makes sense, but it's completely fine if it's not what we want and it needs to be redone. We can take our time reviewing this to make sure we won't need to re-do the library structure again anytime soon. This is why I'm tagging this as a RFC.

Fixes #80647

### CODE REVIEW

File: llvm/utils/TableGen/Basic/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+set(LLVM_LINK_COMPONENTS
+  Support
+  TableGen
+  )
+
+add_llvm_library(LLVMTableGenBasic STATIC OBJECT EXCLUDE_FROM_ALL
+  CodeGenIntrinsics.cpp
+  SDNodeProperties.cpp
+)
+set_target_properties(LLVMTableGenBasic PROPERTIES FOLDER "Tablegenning")
```
Comment (arsenm) [inline llvm/utils/TableGen/Basic/CMakeLists.txt:15]: "Tablegenning"?

**Review by arsenm — COMMENTED** [review]
Most of the diff seems to just clang-format changes

File: llvm/utils/TableGen/Basic/CMakeLists.txt (hunk 1 above)
Comment (Pierre-vh) [reply llvm/utils/TableGen/Basic/CMakeLists.txt:15]: Yeah I just copy pasted that. It seems like we tend to set that properties for tablegen-related stuff. Not sure why

**Comment by Pierre-vh** [comment]
> Most of the diff seems to just clang-format changes

Indeed, it seems like git clang-format formatted the whole files when I moved them.
There's a lot of include changes though, all of the backends include their helpers using `Common/` or `Basic/`. Formatting was needed to sort the includes again

**Review by chapuni — COMMENTED** [review]
Excuse me, it will take longer time for me to review. I am dubious a bit to this direction.

Re. formatting changes, It's good thing to reformat files to be moved. Then I suggest you to split reformatting commits in advance before merging this.

**Comment by Pierre-vh** [comment]
ping

**Comment by Pierre-vh** [comment]
ping :)

**Review by arsenm — APPROVED** [review]

**Comment by Pierre-vh** [comment]
ping

**Review by llvm-beanz — APPROVED** [review]
@Pierre-vh, I think this looks good and @arsenm has approved.

@chapuni is based in Japan so there will be a time lag. Give him another day if he doesn't respond I'd say go ahead and merge this.

File: llvm/utils/TableGen/CMakeLists.txt [hunk 2]
```diff
@@ -1,26 +1,25 @@
...
 
-add_llvm_library(LLVMTableGenCommon STATIC OBJECT EXCLUDE_FROM_ALL
+set(LLVM_LINK_COMPONENTS Support)
+
+# llvm-min-tablegen only contains a subset of backends necessary to
+# build llvm/include. It must not depend on TableGenCommon, as
+# TableGenCommon depends on this already to generate things such as
+# ValueType definitions.
+add_tablegen(llvm-min-tblgen LLVM_HEADERS
+  TableGen.cpp
```
Comment (nico) [inline llvm/utils/TableGen/CMakeLists.txt:15]: Keep this source list alphabetized?

**Review by nico — COMMENTED** [review]
Nice!

An old wish I had back when chapuni introduced llvm-min-tblgen: It'd be nice if llvm-min-tblgen had its own directory, so that we don't need the `PARTIAL_SOURCES_INTENDED` kludge. Looks like this is even easier to do after this change :)

**Pull request merged**